### PR TITLE
Add indexing,search BwC tests for type mapping removal changes

### PR DESCRIPTION
**Pull request by kotwanikunal**
Signed-off-by: Kunal Kotwani <kkotwani@amazon.com>

### Description
- Add backwards compatibility tests for testing type mapping changes from OS2.0
 
### Issues Resolved
- https://github.com/opensearch-project/OpenSearch/issues/2240
 
### Check List
- [ ] New functionality includes testing.
- [x] All tests pass
- [ ] New functionality has been documented.
- [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/IndexingIT.java [hunk 1]
```diff
@@ -202,6 +190,72 @@ public void testAutoIdWithOpTypeCreate() throws IOException {
         }
     }
 
+    public void testTypeRemovalIndexing() throws IOException {
+        final String indexName = "test_index_with_mapping";
+        final String indexWithoutTypeName = "test_index_without_mapping";
+        final String mapping = "\"properties\":{\"f1\":{\"type\":\"keyword\"},\"f2\":{\"type\":\"keyword\"}}";
```
Comment (dreamer-89) [inline qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/IndexingIT.java:196]: I think we want to add type to mappings. E.g. below
```
"{\"_doc\": {\"properties\": {\"field\": {\"type\": \"text\"}}}}"
```

**Review by dreamer-89 — CHANGES_REQUESTED** [review]

**Comment by dreamer-89** [comment]
Gradle check failure. 
```
> Task :qa:rolling-upgrade:v1.3.1#upgradedClusterTest

REPRODUCE WITH: ./gradlew ':qa:rolling-upgrade:v1.3.1#upgradedClusterTest' --tests "org.opensearch.upgrades.IndexingIT.testTypeRemovalIndexing" -Dtests.seed=A00E1BB3E539D943 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=en-PH -Dtests.timezone=Europe/Helsinki -Druntime.java=17

org.opensearch.upgrades.IndexingIT > testTypeRemovalIndexing FAILED
    java.lang.ExceptionInInitializerError
        at __randomizedtesting.SeedInfo.seed([A00E1BB3E539D943:21AD8924DEBE40D]:0)
        at org.opensearch.upgrades.IndexingIT.assertNumHits(IndexingIT.java:307)
        at org.opensearch.upgrades.IndexingIT.testTypeRemovalIndexing(IndexingIT.java:232)

        Caused by:
        java.lang.IllegalArgumentException: Failed to parse value [null] as only [true] or [false] are allowed.
            at org.opensearch.common.Booleans.parseBoolean(Booleans.java:86)
            at org.opensearch.upgrades.AbstractFullClusterRestartTestCase.<clinit>(AbstractFullClusterRestartTestCase.java:45)
            ... 2 more
```

**Comment by dreamer-89** [comment]
Failure doesn't belong to code changes. Re-firing

**Comment by dreamer-89** [comment]
start gradle check

**Comment by kotwanikunal** [comment]
Another unrelated test failure - 
```
REPRODUCE WITH: ./gradlew ':qa:remote-clusters:integTest' --tests "org.opensearch.cluster.remote.test.RemoteClustersIT.testHAProxyModeConnectionWorks" -Dtests.seed=DB8B6E2943623CFB -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=mt-MT -Dtests.timezone=SystemV/AST4 -Druntime.java=17

org.opensearch.cluster.remote.test.RemoteClustersIT > testHAProxyModeConnectionWorks FAILED
    java.lang.AssertionError
        at __randomizedtesting.SeedInfo.seed([DB8B6E2943623CFB:DCFDBBE278E44AA6]:0)
        at org.junit.Assert.fail(Assert.java:87)
        at org.junit.Assert.assertTrue(Assert.java:42)
        at org.junit.Assert.assertTrue(Assert.java:53)
        at org.opensearch.cluster.remote.test.RemoteClustersIT.testHAProxyModeConnectionWorks(RemoteClustersIT.java:125)
```

**Comment by kotwanikunal** [comment]
start gradle check

**Comment by dreamer-89** [comment]
There is already issue opened for `testHAProxyModeConnectionWorks`

https://github.com/opensearch-project/OpenSearch/issues/1703

**Comment by dreamer-89** [comment]
start gradle check

**Comment by kotwanikunal** [comment]
Tests are passing. 
Seeing the following failure - 

```
* What went wrong:
Execution failed for task ':distribution:bwc:bugfix:buildBwcLinuxTar'.
> Building 1.3.1 didn't generate expected file /var/CITOOL/workflow/OpenSearch_CI/PR_Checks/Gradle_Check/search/distribution/bwc/bugfix/build/bwc/checkout-1.3/distribution/archives/linux-tar/build/distributions/opensearch-min-1.3.1-SNAPSHOT-linux-x64.tar.gz

```

**Comment by dreamer-89** [comment]
> Seeing the following failure -
> 
> ```
> * What went wrong:
> Execution failed for task ':distribution:bwc:bugfix:buildBwcLinuxTar'.
> > Building 1.3.1 didn't generate expected file /var/CITOOL/workflow/OpenSearch_CI/PR_Checks/Gradle_Check/search/distribution/bwc/bugfix/build/bwc/checkout-1.3/distribution/archives/linux-tar/build/distributions/opensearch-min-1.3.1-SNAPSHOT-linux-x64.tar.gz
> ```

There was version bump from 1.3.1 -> 1.3.2 [here](https://github.com/opensearch-project/OpenSearch/pull/2675/files). Not sure, why it impacts this PR running against main.

**Comment by dreamer-89** [comment]
start gradle check

**Comment by dreamer-89** [comment]
I think we need to add `1.3.2` in `main` branch as well.

**Comment by dreamer-89** [comment]
start gradle check

**Comment by tlfeng** [comment]
Congratulations that you got the 4000th Gradle check runs in CI 😄 🎉

**Comment by kotwanikunal** [comment]
Error in the logs -
```
REPRODUCE WITH: ./gradlew ':qa:rolling-upgrade:v1.3.2#upgradedClusterTest' --tests "org.opensearch.upgrades.IndexingIT.testTypeRemovalSnapshots" -Dtests.seed=7455B1B271CECBC9 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=sr-Latn-BA -Dtests.timezone=CET -Druntime.java=17

org.opensearch.upgrades.IndexingIT > testTypeRemovalSnapshots FAILED
    org.opensearch.client.ResponseException: method [POST], host [http://[::1]:41251], URI [_snapshot/test_repository/test_snapshot/_restore?wait_for_completion=true], status line [HTTP/1.1 500 Internal Server Error]
    {"error":{"root_cause":[{"type":"snapshot_restore_exception","reason":"[test_repository:test_snapshot] snapshot does not exist"}],"type":"snapshot_restore_exception","reason":"[test_repository:test_snapshot] snapshot does not exist"},"status":500}
        at __randomizedtesting.SeedInfo.seed([7455B1B271CECBC9:A68E597FC445E45E]:0)
        at app//org.opensearch.client.RestClient.convertResponse(RestClient.java:350)
        at app//org.opensearch.client.RestClient.performRequest(RestClient.java:320)
        at app//org.opensearch.client.RestClient.performRequest(RestClient.java:295)
        at app//org.opensearch.test.rest.OpenSearchRestTestCase.restoreSnapshot(OpenSearchRestTestCase.java:1120)
        at app//org.opensearch.upgrades.IndexingIT.testTypeRemovalSnapshots(IndexingIT.java:353)


```

Retrying on my local

**Comment by kotwanikunal** [comment]
Local run works as expected.
Retrying the check
```
dev-dsk-kkotwani % ./gradlew :qa:rolling-upgrade:v1.3.2#upgradedClusterTest -Dtests.class=org.opensearch.upgrades.IndexingIT -Dtests.seed=7455B1B271CECBC9 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=sr-Latn-BA -Dtests.timezone=CET -Druntime.java=17

> Configure project :qa:os
Cannot add task 'destructiveDistroTest.docker' as a task with that name already exists.
=======================================
OpenSearch Build Hamster says Hello!
  Gradle Version        : 7.4.2
  OS Info               : Linux 5.4.181-109.354.amzn2int.x86_64 (amd64)
  Runtime JDK Version   : 17 (Amazon Corretto JDK)
  Runtime java.home     : /home/kkotwani/.sdkman/candidates/java/17.0.1.12.1-amzn
  Gradle JDK Version    : 11 (Amazon Corretto JDK)
  Gradle java.home      : /local/home/kkotwani/.sdkman/candidates/java/11.0.13.8.1-amzn
  Random Testing Seed   : 7455B1B271CECBC9
  In FIPS 140 mode      : false
=======================================

> Task :distribution:bwc:bugfix:checkoutBwcBranch
Performing checkout of opensearch-project/1.3...
Checkout hash for :distribution:bwc:bugfix is 83d51ae2f63edbc4e43b927b208bd6cccd85f788

> Task :qa:rolling-upgrade:v1.3.2#oldClusterTest
OpenJDK 64-Bit Server VM warning: Ignoring option --illegal-access=warn; support was removed in 17.0
WARNING: A terminally deprecated method in java.lang.System has been called
WARNING: System::setSecurityManager has been called by org.opensearch.bootstrap.BootstrapForTesting (file:/local/home/kkotwani/workspace/OpenSearch/test/framework/build/distributions/framework-2.0.0-SNAPSHOT.jar)
WARNING: Please consider reporting this to the maintainers of org.opensearch.bootstrap.BootstrapForTesting
WARNING: System::setSecurityManager will be removed in a future release
WARNING: A terminally deprecated method in java.lang.System has been called
WARNING: System::setSecurityManager has been called by org.gradle.api.internal.tasks.testing.worker.TestWorker (file:/local/home/kkotwani/.gradle/wrapper/dists/gradle-7.4.2-all/9uukhhbclvbegdvsww0j0cr3p/gradle-7.4.2/lib/plugins/gradle-testing-base-7.4.2.jar)
WARNING: Please consider reporting this to the maintainers of org.gradle.api.internal.tasks.testing.worker.TestWorker
WARNING: System::setSecurityManager will be removed in a future release

> Task :qa:rolling-upgrade:v1.3.2#oneThirdUpgradedTest
OpenJDK 64-Bit Server VM warning: Ignoring option --illegal-access=warn; support was removed in 17.0
WARNING: A terminally deprecated method in java.lang.System has been called
WARNING: System::setSecurityManager has been called by org.opensearch.bootstrap.BootstrapForTesting (file:/local/home/kkotwani/workspace/OpenSearch/test/framework/build/distributions/framework-2.0.0-SNAPSHOT.jar)
WARNING: Please consider reporting this to the maintainers of org.opensearch.bootstrap.BootstrapForTesting
WARNING: System::setSecurityManager will be removed in a future release
WARNING: A terminally deprecated method in java.lang.System has been called
WARNING: System::setSecurityManager has been called by org.gradle.api.internal.tasks.testing.worker.TestWorker (file:/local/home/kkotwani/.gradle/wrapper/dists/gradle-7.4.2-all/9uukhhbclvbegdvsww0j0cr3p/gradle-7.4.2/lib/plugins/gradle-testing-base-7.4.2.jar)
WARNING: Please consider reporting this to the maintainers of org.gradle.api.internal.tasks.testing.worker.TestWorker
WARNING: System::setSecurityManager will be removed in a future release

> Task :qa:rolling-upgrade:v1.3.2#twoThirdsUpgradedTest
OpenJDK 64-Bit Server VM warning: Ignoring option --illegal-access=warn; support was removed in 17.0
WARNING: A terminally deprecated method in java.lang.System has been called
WARNING: System::setSecurityManager has been called by org.opensearch.bootstrap.BootstrapForTesting (file:/local/home/kkotwani/workspace/OpenSearch/test/framework/build/distributions/framework-2.0.0-SNAPSHOT.jar)
WARNING: Please consider reporting this to the maintainers of org.opensearch.bootstrap.BootstrapForTesting
WARNING: System::setSecurityManager will be removed in a future release
WARNING: A terminally deprecated method in java.lang.System has been called
WARNING: System::setSecurityManager has been called by org.gradle.api.internal.tasks.testing.worker.TestWorker (file:/local/home/kkotwani/.gradle/wrapper/dists/gradle-7.4.2-all/9uukhhbclvbegdvsww0j0cr3p/gradle-7.4.2/lib/plugins/gradle-testing-base-7.4.2.jar)
WARNING: Please consider reporting this to the maintainers of org.gradle.api.internal.tasks.testing.worker.TestWorker
WARNING: System::setSecurityManager will be removed in a future release

> Task :qa:rolling-upgrade:v1.3.2#upgradedClusterTest
OpenJDK 64-Bit Server VM warning: Ignoring option --illegal-access=warn; support was removed in 17.0
WARNING: A terminally deprecated method in java.lang.System has been called
WARNING: System::setSecurityManager has been called by org.opensearch.bootstrap.BootstrapForTesting (file:/local/home/kkotwani/workspace/OpenSearch/test/framework/build/distributions/framework-2.0.0-SNAPSHOT.jar)
WARNING: Please consider reporting this to the maintainers of org.opensearch.bootstrap.BootstrapForTesting
WARNING: System::setSecurityManager will be removed in a future release
WARNING: A terminally deprecated method in java.lang.System has been called
WARNING: System::setSecurityManager has been called by org.gradle.api.internal.tasks.testing.worker.TestWorker (file:/local/home/kkotwani/.gradle/wrapper/dists/gradle-7.4.2-all/9uukhhbclvbegdvsww0j0cr3p/gradle-7.4.2/lib/plugins/gradle-testing-base-7.4.2.jar)
WARNING: Please consider reporting this to the maintainers of org.gradle.api.internal.tasks.testing.worker.TestWorker
WARNING: System::setSecurityManager will be removed in a future release

BUILD SUCCESSFUL in 1m 16s
182 actionable tasks: 10 executed, 172 up-to-date
```

**Comment by kotwanikunal** [comment]
start gradle check

File: qa/rolling-upgrade/build.gradle [hunk 2]
```diff
@@ -60,7 +60,7 @@ for (Version bwcVersion : BuildParams.bwcVersions.wireCompatible) {
       numberOfNodes = 3
 
       setting 'repositories.url.allowed_urls', 'http://snapshot.test*'
-      setting 'path.repo', "${buildDir}/cluster/shared/repo/${baseName}"
+      setting 'path.repo', "['${buildDir}/cluster/shared/repo/${baseName}', '${buildDir}/cluster/shared/repo/${baseName}/test']"
```
Comment (dreamer-89) [inline qa/rolling-upgrade/build.gradle:63]: Why this change is needed ?

File: qa/rolling-upgrade/build.gradle [hunk 3]
```diff
@@ -74,6 +74,7 @@ for (Version bwcVersion : BuildParams.bwcVersions.wireCompatible) {
     }
     systemProperty 'tests.upgrade_from_version', bwcVersionStr
     systemProperty 'tests.rest.suite', 'old_cluster'
+    systemProperty 'tests.repo_location', "${buildDir}/cluster/shared/repo/${baseName}/test"
```
Comment (dreamer-89) [inline qa/rolling-upgrade/build.gradle:77]: Do we need to cleanup repository directory like we are doing for `path.repo` in doFirst block

Comment (kotwanikunal) [reply qa/rolling-upgrade/build.gradle:77]: The cleanup cleans the entire folder structure. I am utilizing a subfolder within the actual repo which should be cleaned up by the existing code.

**Comment by kotwanikunal** [comment]
Unrelated test failure:

```
REPRODUCE WITH: ./gradlew ':plugins:repository-azure:internalClusterTest' --tests "org.opensearch.repositories.azure.AzureBlobStoreRepositoryTests.testSnapshotWithLargeSegmentFiles" -Dtests.seed=A319369D69EB94AA -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=es-DO -Dtests.timezone=America/Fort_Nelson -Druntime.java=17

org.opensearch.repositories.azure.AzureBlobStoreRepositoryTests > testSnapshotWithLargeSegmentFiles FAILED
    java.lang.AssertionError: 
    Expected: a value greater than <0>
         but: <0> was equal to <0>
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:964)
        at org.junit.Assert.assertThat(Assert.java:930)
        at org.opensearch.repositories.blobstore.OpenSearchBlobStoreRepositoryIntegTestCase.assertSuccessfulRestore(OpenSearchBlobStoreRepositoryIntegTestCase.java:532)
        at org.opensearch.repositories.blobstore.OpenSearchBlobStoreRepositoryIntegTestCase.assertSuccessfulRestore(OpenSearchBlobStoreRepositoryIntegTestCase.java:528)
        at org.opensearch.repositories.blobstore.OpenSearchMockAPIBasedRepositoryIntegTestCase.testSnapshotWithLargeSegmentFiles(OpenSearchMockAPIBasedRepositoryIntegTestCase.java:182)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:938)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:974)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:988)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:44)
        at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
        at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
        at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
        at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:817)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:468)
        at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:947)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:832)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:883)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:894)
        at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
        at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
        at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
        at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
        at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
        at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
        at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
        at java.base/java.lang.Thread.run(Thread.java:833
```

**Comment by kotwanikunal** [comment]
start gradle check

**Comment by kotwanikunal** [comment]
Will re run tests after https://github.com/opensearch-project/OpenSearch/pull/2795 gets merged in.

File: qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/IndexingIT.java [hunk 4]
```diff
@@ -202,24 +193,297 @@ public void testAutoIdWithOpTypeCreate() throws IOException {
...
+        final String templateName = "test_template";
+        final String indexNamePattern = "template_test*";
+
+        final String mapping = "\"properties\":{\"f1\":{\"type\":\"keyword\"},\"f2\":{\"type\":\"keyword\"}}";
+        final String typeMapping = "\"_doc\":{" + mapping +"}";
+
+        switch (CLUSTER_TYPE) {
+            case OLD:
+                Version minNodeVersion = getMinNodeVersion();
+                if (minNodeVersion.before(Version.V_2_0_0)) {
```
Comment (tlfeng) [inline qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/IndexingIT.java:209]: When all the nodes in the cluster is in the same version, I think it's easier to directly use the variable `UPGRADE_FROM_VERSION`. (Please see the other test classes in the directory https://github.com/opensearch-project/OpenSearch/blob/1.3.1/qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/JodaCompatibilityIT.java#L77 and https://github.com/opensearch-project/OpenSearch/blob/1.3.1/qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/MappingIT.java#L45)

In addition, I think getting the `minNodeVersion` is originally designed to be used in a mixed-version cluster (`case MIXED`), when an API response from nodes of different versions are different. My conclusion is derived from the original test case `testAutoIdWithOpTypeCreate()` (https://github.com/opensearch-project/OpenSearch/blob/1.3.1/qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/IndexingIT.java#L211-L213)

Further, learning from the above 2 tests classes, probably it will be more concise to use `assumeTrue` to filter the versions to run the test. 😁
```
assumeTrue("Mapping types is removed in 2.0", UPGRADE_FROM_VERSION.beforeVersion.V_2_0_0));
```

File: qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/IndexingIT.java [hunk 5]
```diff
@@ -63,42 +69,42 @@
```
Comment (tlfeng) [inline qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/IndexingIT.java:69]: A small question. Would you consider putting the test cases about mapping type removal to a separate class?
1. There is a new test case to "verify older indices are getting snapshots successfully in OpenSearch 2.0", it seems out of the range of this class "Basic test that indexed documents survive the rolling restart."
2. All the new test cases is only valid when upgrading from version < 2.0, it might be more neat to treat these test in a separate group.

**Review by tlfeng — COMMENTED** [review]
Thank you! It's a great opportunity for me to understand the rolling upgrade test from your new codes. 👍👍

File: qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/IndexingIT.java (hunk 5 above)
Comment (kotwanikunal) [reply qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/IndexingIT.java:62]: The tests are now moved into a `TypeRemovalIT` class

File: qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/IndexingIT.java (hunk 4 above)
Comment (kotwanikunal) [reply qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/IndexingIT.java]: I was leaning on the `minNodeVersion` check to avoid any race/weird condition before I actually push in the typed indices. It does not seem to be an issue in my tests. Moved it to a regular check.

`assertTrue` will not work for all the cases since the `bwcTest` runs checks between not just current branch but a bunch of versions depending on compatibility. (https://github.com/opensearch-project/OpenSearch/blob/1b571ece28f5644becf97c165de853943af73ac5/qa/rolling-upgrade/build.gradle#L40) Backporting to older branches will break the assumption.

File: qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/IndexingIT.java (hunk 5 above)
Comment (tlfeng) [reply qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/IndexingIT.java:62]: Thank you for this structure change! 👍

File: qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/IndexingIT.java (hunk 4 above)
Comment (tlfeng) [reply qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/IndexingIT.java]: Ah, got your reason to use `minNodeVersion` check 😄 👍. The current code looks good, but I need some time to understand the limitation of `assumeTrue` you pointed out.

Comment (tlfeng) [reply qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/IndexingIT.java]: The meaning of `assumeTrue("...", UPGRADE_FROM_VERSION.beforeVersion.V_2_0_0);` is to filter the test to only run when upgrading from a cluster < 2.0 version.
I just learnt this from seeing the other test class, such as https://github.com/opensearch-project/OpenSearch/blob/1.3.1/qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/JodaCompatibilityIT.java#L77

You could see the javadoc of `assumeTrue()` (https://javadoc.io/doc/com.carrotsearch.randomizedtesting/randomizedtesting-runner/2.7.1/com/carrotsearch/randomizedtesting/RandomizedTest.html#assumeTrue-java.lang.String-boolean-)
> condition - If false an AssumptionViolatedException is thrown by this method and the test case (should be) ignored. Tests that are assumption-failures do not break builds.

In my opinion, breaking the assumption is the result we want. 😄

**Review by tlfeng — APPROVED** [review]

File: qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/IndexingIT.java (hunk 4 above)
Comment (kotwanikunal) [reply qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/IndexingIT.java]: I misread it as `assertTrue`! This is a great find. Thanks for linking this 🙂 
The if conditions currently do cover all of over scenarios, but will update with `assumeTrue` whenever I push any changes out for the test.

Comment (tlfeng) [reply qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/IndexingIT.java]: Sounds good! Yeah.. your PR does a good opportunity for me to find those interesting usages.

File: qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/TypeRemovalIT.java [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+import org.opensearch.rest.RestStatus;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Map;
+import java.util.regex.Pattern;
+
+import static org.opensearch.rest.action.search.RestSearchAction.TOTAL_HITS_AS_INT_PARAM;
+
+public class TypeRemovalIT extends AbstractRollingTestCase {
```
Comment (dreamer-89) [inline qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/TypeRemovalIT.java:31]: minor: `TypeRemovalIT` -> `MappingTypeRemovalIT` ? Type is overloaded term in OpenSearch and used in other places.

File: qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/TypeRemovalIT.java [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+        final String templateIndexName = "template_test_index";
+        final String templateName = "test_template";
+        final String indexNamePattern = "template_test*";
+
+        final String mapping = "\"properties\":{\"f1\":{\"type\":\"keyword\"},\"f2\":{\"type\":\"keyword\"}}";
+        final String typeMapping = "\"_doc\":{" + mapping +"}";
+
+        switch (CLUSTER_TYPE) {
+            case OLD:
+                if (UPGRADE_FROM_VERSION.before(Version.V_2_0_0)) {
```
[the remark this answers is not in the document]
Comment (kotwanikunal) [reply qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/TypeRemovalIT.java:45]: Against 3.0, it will run against the previous wire compatible versions (v2.x.x). And yes, it will be a NoOp in the versions ahead.
Let me branch it off 2.0 instead.

[the remark this answers is not in the document]
Comment (kotwanikunal) [reply qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/TypeRemovalIT.java:45]: https://github.com/opensearch-project/OpenSearch/pull/2901

[the remark this answers is not in the document]
Comment (tlfeng) [reply qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/TypeRemovalIT.java:45]: Looks like we also need this test in `2.x` branch? 😂

File: qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/TypeRemovalIT.java [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+        }
+        assertEquals(expectedResponse,
+            EntityUtils.toString(searchTestIndexResponse.getEntity(), StandardCharsets.UTF_8));
+    }
+
+    private final Pattern TYPE_REMOVAL_WARNING = Pattern.compile(
+        "^\\[types removal\\] (.+) include_type_name (.+) is deprecated\\. The parameter will be removed in the next major version\\.$"
+    );
+
+    private void useIgnoreTypesRemovalWarningsHandler(Request request) {
```
Comment (tlfeng) [inline qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/TypeRemovalIT.java:283]: I'm curious is this new method having a same function with the existing method https://github.com/opensearch-project/OpenSearch/blob/1.3.1/test/framework/src/main/java/org/opensearch/test/rest/OpenSearchRestTestCase.java#L326 ?

**Comment by kotwanikunal** [comment]
Closing this in favor of https://github.com/opensearch-project/OpenSearch/pull/2901
